src/node/object_path.cc

```diff
@@ -0,0 +1,322 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <tvm/node/object_path.h>
+#include <tvm/node/repr_printer.h>
+#include <tvm/runtime/memory.h>
+#include <tvm/runtime/registry.h>
+
+#include <algorithm>
+#include <cstring>
+
+using namespace tvm::runtime;
+
+namespace tvm {
+
+// ============== ObjectPathNode ==============
+
+ObjectPathNode::ObjectPathNode(ObjectPathNode* parent)
+    : parent_(GetRef<ObjectRef>(parent)), length_(parent == nullptr ? 1 : parent->length_ + 1) {}
+
+// --- GetParent ---
+
+ObjectPath ObjectPathNode::GetParent() const { return Downcast<ObjectPath>(parent_); }
+
+TVM_REGISTER_GLOBAL("node.ObjectPathGetParent").set_body_typed([](const ObjectPath& path) {
+  return path->GetParent();
+});
+
+// --- Length ---
+
+size_t ObjectPathNode::Length() const { return length_; }
+
+TVM_REGISTER_GLOBAL("node.ObjectPathLength").set_body_typed([](const ObjectPath& path) {
+  return static_cast<int64_t>(path->Length());
+});
+
+// --- GetPrefix ---
+
+ObjectPath ObjectPathNode::GetPrefix(size_t length) const {
+  if (length > Length()) {
+    throw std::out_of_range("Attempted to get a prefix longer than the path itself");
+  }
+
+  const ObjectPathNode* node = this;
+  size_t suffix_len = Length() - length;
+  for (size_t i = 0; i < suffix_len; ++i) {
+    node = node->ParentNode();
+  }
+
+  return GetRef<ObjectPath>(node);
+}
+
+TVM_REGISTER_GLOBAL("node.ObjectPathGetPrefix")
+    .set_body_typed([](const ObjectPath& path, int64_t length) {
+      if (length < 0) {
+        throw std::out_of_range("Prefix length can't be negative");
+      }
+      return path->GetPrefix(static_cast<size_t>(length));
```

Comment:
Just I would love to discuss two things here, mostly about styles:

First, besides legacy in the codebase, in new APIs, we strive to make sure the C++ and Python API are consistent. Therefore, usually we do not prefer putting logic in registration to prevent potential discrepancy.

The second is about the usage of `size_t`. It's more like personal preference so I dont have strong opinion: usually I would prefer `int64_t`/`int32_t` to `size_t` if there is no specific need (e.g. large uint, compatibility with STL), because unsigned types could cause some arithmetic issues and `size_t` could be platform-dependent (IIRC).

Therefore, how about we prefer using `int` or `int64_t` throughout `ObjectPath` work if there is no specific need? I would imagine it simplify the implementation and eliminate API discrepancy.

Reply:
Sounds good, done